sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala

```diff
@@ -192,6 +192,23 @@ trait HigherOrderFunction extends Expression with ExpectsInputTypes {
           argumentMap(variable.exprId)
       }
   }
+
+  override lazy val canonicalized: Expression = {
+    var currExprId = -1
+    val argumentMap = functions.flatMap(_.collect {
+      case l: NamedLambdaVariable =>
+        currExprId += 1
+        l.name -> currExprId
```

Comment:
Are you sure the name is unique? It looks safer to use `l.exprId` as the key.

Reply:
After #32424, I think it should be unique, but I agree that `l.exprId` is safer.